synapse/handlers/oidc_handler.py

```diff
@@ -854,6 +867,15 @@ async def handle_oidc_callback(
             )
 
         # otherwise, it's a login
+        logger.debug("Userinfo for OIDC login: %s", userinfo)
+
+        # Ensure that the attributes of the logged in user meet the required
+        # attributes
+        # With OIDC, we check userinfo against attribute_requirements
+        if not self._sso_handler.check_required_attributes(
+            request, userinfo, self._oidc_attribute_requirements
+        ):
```

Comment:
I'm not sure this is quite the same, `check_required_attributes` expects the attributes to be a mapping of attribute names (as strings) to a list of values, I *think* `userinfo` is a mapping of attribute name to a single value.

I think the code around https://github.com/matrix-org/synapse/blob/df425c2c63d969778f725c9053dd93e57ee854e1/synapse/handlers/sso.py#L977-L979 will break pretty badly if you do things like require a value of `foo` and the actual value is `foobar`.

I think the solution is to pass in `{k: [v] for k, v in userinfo.items()}` to `check_required_attributes`.

Reply:
Yeah. The behavior here gets a bit odd - userinfo here is a dict, with keys being strings and values being parsed from JSON data types to python data types - so it's a mix of strings, booleans, integers, lists, and whatever other arbitrary elements have been added to the scopes for the oidc provider. 

Doing `{k: [v] for k, v in userinfo.items()}` would ensure that things like strings are handled as exact matches, but I'm unsure how `if req.value in values` acts if `values` is a list containing a list when `req.value` is only ever going to be a string, since that's what `SsoRequirement.JSON_SCHEMA` demands.

Ideally, I'd like it to support behavior like "userinfo groups contains `foobar` when groups is a list of strings", but I'm not sure how to cleanly handle that.

As it is right now, it's always a loose match - `values` here will be an integer, a string, a boolean, or a list, and `req.value` will be a string. If `values` is a string and `req.value` is a substring of it, the check passes. If `values` is a list, and it contains the string that `req.value` is, the check passes. In all other cases, the check fails.
This is perhaps okay, but it is a bit sloppy, and the documentation as written doesn't explain it very well.

Perhaps I could expand `SsoRequirement.JSON_SCHEMA` such that `req.value` here can be things other than a string, and add an `operator` field to it to support both the exact and contains style of matching? That would require changes to the CAS and SAML handlers as well though.
But that would allow `attribute_requirements` to support both the "attribute must match X" behavior it currently does in SAML and CAS, and a new "attribute must contain X" behavior, which would be ideal for things like checking group membership.
`groups` is a fairly common OIDC userinfo attribute, after all.

Reply:
Would groups be returned as a string or a list of strings? (Or is that up to the implementation to decide?)

I'll need to think a bit more about this.

Reply:
It should be up to the OIDC provider - as far as I'm aware the OpenID Connect spec only requires that the userinfo response is valid JSON, and that it must contain a `sub` value, and that it must not contain any empty strings or null values.

I believe most OIDC providers handle it as a list of strings, though.